### PR TITLE
[BUGFIX] Do not sort by key when receiving paths arrays

### PR DESCRIPTION
**Pull request by NamelessCoder**
During the decoupling a decision was made to not use
array keys for the priority of template paths - instead,
the order of the array should determine the lookup order.

However, the TemplatePaths class of TYPO3Fluid had
received a sorting behavior that was intended for the
TemplatePaths class of TYPO3 CMS. This patch removes
that sorting behavior.

This fix must be followed up by a patch for TYPO3 CMS.
suggested strategy there is to override the setter methods
for paths and enforce the sorting there.

References: #104
References: https://forge.typo3.org/issues/75862
References: https://review.typo3.org/#/c/48279/

### CODE REVIEW

**Comment by helhum** [comment]
no test change needed for this? 😧

File: src/View/TemplatePaths.php [hunk 1]
```diff
@@ -666,7 +655,7 @@ protected function getPartialPathAndFilename($partialName) {
 	 */
 	protected function resolveFileInPaths(array $paths, $relativePathAndFilename, $format = self::DEFAULT_FORMAT) {
 		$tried = array();
-		foreach ($paths as $path) {
+		while ($path = array_pop($paths)) {
```
Comment (helhum) [inline src/View/TemplatePaths.php]: what is the point of this change?

File: src/View/TemplatePaths.php [hunk 2]
```diff
@@ -241,17 +241,11 @@ public function resolveTemplateFileForControllerAndActionAndFormat($controller,
...
-				$candidate = $this->ensureAbsolutePath($templateRootPath . $identifier);
-				if (file_exists($candidate)) {
-					return self::$resolvedFiles['templates'][$identifier] = $candidate;
-				}
+			$templateRootPaths = $this->getTemplateRootPaths();
+			try {
+				return self::$resolvedFiles['templates'][$identifier] = $this->resolveFileInPaths($templateRootPaths, $controller . '/' . $action, $format);
+			} catch (InvalidTemplateResourceException $error) {
+				return self::$resolvedFiles['templates'][$identifier] = $this->resolveFileInPaths($templateRootPaths, $action, $format);
 			}
```
Comment (helhum) [inline src/View/TemplatePaths.php]: I do not see a connection between nor sorting paths and this change. What is this change good for?

File: src/View/TemplatePaths.php (hunk 1 above)
Comment (NamelessCoder) [inline src/View/TemplatePaths.php]: Reversing the order of iteration

File: src/View/TemplatePaths.php (hunk 2 above)
Comment (NamelessCoder) [inline src/View/TemplatePaths.php]: Refactoring to call existing method to resolve file - that method has now been fixed to reverse the iteration order (see above)

**Comment by NamelessCoder** [comment]
No test change required. Test cases were already providing and expecting the right orders (since this krsort behavior was never intended to be in Fluid itself)

File: src/View/TemplatePaths.php (hunk 1 above)
Comment (helhum) [inline src/View/TemplatePaths.php]: ah, of course. looks a bit non obvious, but I could live with that

Comment (NamelessCoder) [inline src/View/TemplatePaths.php]: You're right - I'll put this as a code comment for anyone reading there and not reading this pull request.

File: src/View/TemplatePaths.php [hunk 3]
```diff
@@ -241,17 +241,11 @@ public function resolveTemplateFileForControllerAndActionAndFormat($controller,
...
-			foreach ($this->getTemplateRootPaths() as $templateRootPath) {
-				$candidate = $this->ensureAbsolutePath($templateRootPath . $identifier);
-				if (file_exists($candidate)) {
-					return self::$resolvedFiles['templates'][$identifier] = $candidate;
-				}
+			$templateRootPaths = $this->getTemplateRootPaths();
+			try {
+				return self::$resolvedFiles['templates'][$identifier] = $this->resolveFileInPaths($templateRootPaths, $controller . '/' . $action, $format);
+			} catch (InvalidTemplateResourceException $error) {
+				return self::$resolvedFiles['templates'][$identifier] = $this->resolveFileInPaths($templateRootPaths, $action, $format);
```
Comment (helhum) [inline src/View/TemplatePaths.php]: not sure, but don't we potentially then throw an exception here which was not thrown before?

Comment (NamelessCoder) [inline src/View/TemplatePaths.php]: Ah, yes and no... `getTemplateSource` will throw the same exception if this method returns NULL or the file does not exist - but the exception thrown from that method is a LOT more descriptive. Will refactor!

**Pull request merged**